datafusion-examples/examples/table_sample.rs

```diff
@@ -0,0 +1,1353 @@
+                        return plan_err!(
+                            "quantity must be a number: {:?}",
+                            quantity.value
+                        );
+                    }
+                    let value = value.unwrap() / 100.0;
+                    Self::new_node(input, value, None, seed)
+                }
+                None => {
+                    // Clickhouse-style sample
+                    let value = evaluate_number::<f64>(&quantity.value);
+                    if value.is_none() {
+                        return plan_err!(
+                            "quantity must be a valid number: {:?}",
+                            quantity.value
+                        );
+                    }
+                    let value = value.unwrap();
+                    if value < 0.0 {
+                        return plan_err!(
+                            "quantity must be a non-negative number: {:?}",
+                            quantity.value
+                        );
+                    }
+                    if value >= 1.0 {
+                        // If value is larger than 1, it is a row limit
+                        LogicalPlanBuilder::from(input)
+                            .limit(0, Some(value as usize))?
+                            .build()
+                    } else {
+                        // If value is between 0.0 and 1.0, it is a fraction
+                        Self::new_node(input, value, None, seed)
+                    }
+                }
+            };
+        }
+        plan_err!("Cannot plan sample SQL")
+    }
+
+    fn create_logical_plan(&self, statement: Statement) -> Result<LogicalPlan> {
+        let sql_to_rel = SqlToRel::new(self.context_provider);
+
+        let stmt = statement.clone();
+        match &stmt {
+            Statement::Query(query) => {
+                let inner_plan = sql_to_rel.sql_statement_to_plan(statement)?;
+                if let SetExpr::Select(select) = &*query.body {
+                    if select.from.len() == 1 {
+                        let table_with_joins = select.from.first().unwrap();
+                        if let TableFactor::Table {
+                            sample: Some(table_sample_kind),
+                            ..
+                        } = &table_with_joins.relation
+                        {
+                            debug!("Constructing table sample plan from {:?}", &select);
+                            return self.sample_to_logical_plan(
+                                inner_plan,
+                                table_sample_kind.clone(),
+                            );
+                        }
+                    }
+                }
+                // Pass-through by default
+                Ok(inner_plan)
+            }
+            _ => sql_to_rel.sql_statement_to_plan(statement),
+        }
+    }
+}
+
+// Context provider for tests
+
+struct MockContextProvider<'a> {
```

Comment:
What do you think about  making `SessionContextProvider` public so it can be more easily reused? Having to define a very similar one here is a fairly large usability papercut in my opinion

https://github.com/apache/datafusion/blob/0181c7906b47f9b17fbde92bf30b2d939a80230c/datafusion/core/src/execution/session_state.rs#L1667-L1666